core/peerset/src/peersstate.rs

```diff
@@ -0,0 +1,479 @@
+	In,
+	/// We are connected through an outgoing connection.
+	Out,
+	/// We are not connected to this node.
+	NotConnected,
+}
+
+impl ConnectionState {
+	/// Returns `true` for `In` and `Out`.
+	fn is_connected(self) -> bool {
+		match self {
+			ConnectionState::In => true,
+			ConnectionState::Out => true,
+			ConnectionState::NotConnected => false,
+		}
+	}
+}
+
+impl PeersState {
+	/// Builds a new empty `PeersState`.
+	pub fn new(in_peers: u32, out_peers: u32) -> Self {
+		PeersState {
+			nodes: HashMap::new(),
+			max_in: in_peers,
+			max_out: out_peers,
+		}
+	}
+
+	/// Adds `value` to the reputation of all the nodes we are connected to.
+	///
+	/// In case of overflow, the value of capped.
+	pub fn connected_reputation_increase(&mut self, value: i32) {
+		for (_, peer_state) in self.nodes.iter_mut() {
+			if peer_state.connection_state.is_connected() {
+				peer_state.reputation = peer_state.reputation.saturating_add(value);
+			}
+		}
+	}
+
+	/// Returns an object that grants access to the state of a peer.
+	pub fn peer<'a>(&'a mut self, peer_id: &'a PeerId) -> Peer<'a> {
+		if let Some(node) = self.nodes.get(peer_id) {
+			if node.connection_state.is_connected() {
+				Peer::Connected(ConnectedPeer {
+					parent: self,
+					peer_id: Cow::Borrowed(peer_id),
+				})
+			} else {
+				Peer::NotConnected(NotConnectedPeer {
+					parent: self,
+					peer_id: Cow::Borrowed(peer_id),
+				})
+			}
+		} else {
+			Peer::Unknown(UnknownPeer {
+				parent: self,
+				peer_id: Cow::Borrowed(peer_id),
+			})
+		}
+	}
+
+	/// Returns the list of all the peers we know of.
+	// Note: this method could theoretically return a `Peer`, but implementing that
+	// isn't simple.
+	pub fn peers(&self) -> impl Iterator<Item = &PeerId> {
+		self.nodes.keys()
+	}
+
+	/// Returns the list of peers we are connected to.
+	// Note: this method could theoretically return a `ConnectedPeer`, but implementing that
+	// isn't simple.
+	pub fn connected_peers(&self) -> impl Iterator<Item = &PeerId> {
+		self.nodes.iter()
+			.filter(|(_, p)| p.connection_state.is_connected())
+			.map(|(p, _)| p)
+	}
+
+	/// Returns the first reserved peer that we are not connected to.
+	///
+	/// If multiple nodes are reserved, which one is returned is unspecified.
+	pub fn reserved_not_connected_peer(&mut self) -> Option<NotConnectedPeer> {
+		let peer_id = self.nodes.iter_mut()
+			.find(|(_, &mut Node { connection_state, reserved, .. })| {
+				reserved && !connection_state.is_connected()
+			})
+			.map(|(peer_id, _)| peer_id.clone());
+
+		if let Some(peer_id) = peer_id {
+			Some(NotConnectedPeer {
+				parent: self,
+				peer_id: Cow::Owned(peer_id),
+			})
+		} else {
+			None
+		}
+	}
+
+	/// Returns the peer with the highest reputation and that we are not connected to.
+	///
+	/// If multiple nodes have the same reputation, which one is returned is unspecified.
+	pub fn highest_not_connected_peer(&mut self) -> Option<NotConnectedPeer> {
```

Comment:
If this function will be called often, we should use a data structure such as a priority queue or Fibonacci heap.